sql/src/main/java/io/crate/planner/operators/HashJoin.java

```diff
@@ -22,42 +22,182 @@
 import io.crate.expression.symbol.SelectSymbol;
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.Symbols;
 import io.crate.planner.ExecutionPlan;
 import io.crate.planner.PlannerContext;
+import io.crate.planner.ResultDescription;
+import io.crate.planner.distribution.DistributionInfo;
+import io.crate.planner.node.dql.join.Join;
+import io.crate.planner.node.dql.join.JoinType;
 
 import javax.annotation.Nullable;
 import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
+import static io.crate.planner.operators.LogicalPlanner.NO_LIMIT;
+
 class HashJoin extends TwoInputPlan {
 
+    private final Symbol joinCondition;
+    @VisibleForTesting
+    AnalyzedRelation leftRelation;
+    @VisibleForTesting
+    AnalyzedRelation rightRelation;
+
     HashJoin(LogicalPlan lhs,
-             LogicalPlan rhs) {
+             LogicalPlan rhs,
+             Symbol joinCondition,
+             AnalyzedRelation leftRelation,
+             AnalyzedRelation rightRelation) {
         super(lhs, rhs, new ArrayList<>());
+        this.leftRelation = leftRelation;
+        this.rightRelation = rightRelation;
+        this.joinCondition = joinCondition;
+        this.outputs.addAll(lhs.outputs());
+        this.outputs.addAll(rhs.outputs());
+    }
+
+    public JoinType joinType() {
+        return JoinType.INNER;
     }
 
+    protected Symbol joinCondition() {
+        return joinCondition;
+    }
+
+    public Map<LogicalPlan, SelectSymbol> dependencies() {
+        HashMap<LogicalPlan, SelectSymbol> deps = new HashMap<>(lhs.dependencies().size() + rhs.dependencies().size());
+        deps.putAll(lhs.dependencies());
+        deps.putAll(rhs.dependencies());
+        return deps;
+    }
 
     @Override
-    public ExecutionPlan build(PlannerContext plannerContext, ProjectionBuilder projectionBuilder, int limit, int offset, @Nullable OrderBy order, @Nullable Integer pageSizeHint, Row params, Map<SelectSymbol, Object> subQueryValues) {
-        throw new UnsupportedOperationException("HashJoin not implemented");
+    public ExecutionPlan build(PlannerContext plannerContext,
+                               ProjectionBuilder projectionBuilder,
+                               int limit,
+                               int offset,
+                               @Nullable OrderBy order,
+                               @Nullable Integer pageSizeHint,
+                               Row params,
+                               Map<SelectSymbol, Object> subQueryValues) {
+
+        ExecutionPlan left = lhs.build(
+            plannerContext, projectionBuilder, NO_LIMIT, 0, null, null, params, subQueryValues);
+        ExecutionPlan right = rhs.build(
+            plannerContext, projectionBuilder, NO_LIMIT, 0, null, null, params, subQueryValues);
+
+        // Move smaller table to the left side
```

Comment:
should we maybe have a method instead of this comment?

Reply:
the method would be weird... Would have to return a `Tuple<ExecutionPlan, ExecutionPlan>` as `left` and `right` are local variables...